mlir/lib/Analysis/Presburger/Simplex.cpp

```diff
@@ -0,0 +1,919 @@
+Simplex::findPivot(int row, Direction direction) const {
+  llvm::Optional<unsigned> col;
+  for (unsigned j = 2; j < nCol; ++j) {
+    int64_t elem = tableau(row, j);
+    if (elem == 0)
+      continue;
+
+    if (unknownFromColumn(j).restricted &&
+        !signMatchesDirection(elem, direction))
+      continue;
+    if (!col || colVar[j] < colVar[*col])
+      col = j;
+  }
+
+  if (!col)
+    return {};
+
+  Direction newDirection =
+      tableau(row, *col) < 0 ? flippedDirection(direction) : direction;
+  auto opt = findPivotRow(row, newDirection, *col);
+  return std::pair<unsigned, unsigned>{opt.getValueOr(row), *col};
+}
+
+void Simplex::swapRowWithCol(unsigned row, unsigned col) {
+  std::swap(rowVar[row], colVar[col]);
+  Unknown *uCol = &unknownFromColumn(col);
+  Unknown *uRow = &unknownFromRow(row);
+  uCol->ownsRow = false;
+  uRow->ownsRow = true;
+  uCol->pos = col;
+  uRow->pos = row;
+}
+
+void Simplex::pivot(const std::pair<unsigned, unsigned> &p) {
+  pivot(p.first, p.second);
+}
+
+// Pivot pivotRow and pivotCol.
+//
+// Let R be the pivot row unknown and let C be the pivot col unknown.
+// Since initially R = a*C + sum b_i * X_i
+// (where the sum is over the other column's unknowns, x_i)
+// C = (R - (sum b_i * X_i))/a
+//
+// Let u be some other row unknown.
+// u = c*C + sum d_i * X_i
+// So u = c*(R - sum b_i * X_i)/a + sum d_i * X_i
+//
+// This results in the following transform:
+//            pivot col    other col                   pivot col    other col
+// pivot row     a             b       ->   pivot row     1/a         -b/a
+// other row     c             d            other row     c/a        d - bc/a
+//
+// Taking into account the common denominators p and q:
+//
+//            pivot col    other col                    pivot col   other col
+// pivot row     a/p          b/p     ->   pivot row      p/a         -b/a
+// other row     c/q          d/q          other row     cp/aq    (da - bc)/aq
+//
+// The pivot row transform is accomplished be swapping a with the pivot row's
+// common denominator and negating the pivot row except for the pivot column
+// element.
+void Simplex::pivot(unsigned pivotRow, unsigned pivotCol) {
+  assert(pivotCol >= 2 && "Refusing to pivot invalid column");
+
+  swapRowWithCol(pivotRow, pivotCol);
+  std::swap(tableau(pivotRow, 0), tableau(pivotRow, pivotCol));
+  // We need to negate the whole pivot row except for the pivot column.
+  if (tableau(pivotRow, 0) < 0) {
+    // If the denominator is negative, we negate the row by simply negating the
+    // denominator.
+    tableau(pivotRow, 0) = -tableau(pivotRow, 0);
+    tableau(pivotRow, pivotCol) = -tableau(pivotRow, pivotCol);
+  } else {
+    for (unsigned col = 1; col < nCol; ++col) {
+      if (col == pivotCol)
+        continue;
+      tableau(pivotRow, col) = -tableau(pivotRow, col);
+    }
+  }
+  normalizeRow(pivotRow);
+
+  for (unsigned row = 0; row < nRow; ++row) {
+    if (row == pivotRow)
```

Comment:
can you or with the next condition? i.e., if ((row == pivotRow) || (tableau(/* */)) continue;

Reply:
Hm, I wrote it like this since that looks closer to the snippets in the coding standards: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Reply:
This is interesting. Not sure if this is followed in MLIR. Ignore my comment.